modules/hal_nxp/CMakeLists.txt

```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) 2021 Linaro, Limited
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+add_subdirectory(${ZEPHYR_CURRENT_MODULE_DIR} hal_nxp)
```

Comment:
Isn't this a bit of a cross-dependency? You are now including `hal_nxp/` from `zephyr/modules/hal_nxp`, but hal_nxp/zephyr is the one that triggered the inclusion of `zephyr/modules/hal_nxp`.

Also, why do you specify the binary dir?

Reply:
Open to suggestions on how to do this?

I needed the binary dir to address the following cmake issue w/o it:
```
CMake Error at ../../modules/hal_nxp/CMakeLists.txt:7 (add_subdirectory):
  add_subdirectory not given a binary directory but the given source
  directory "/home/galak/git/modules/hal/nxp" is not a subdirectory of
  "/home/galak/git/zephyr/modules/hal_nxp".  When specifying an out-of-tree
  source a binary directory must be explicitly specified.
```

Reply:
>Open to suggestions on how to do this?

You move `hal_nxp/CMakeLists.txt` to `zephyr/modules/hal_nxp/`

See [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_nordic/CMakeLists.txt)

Reply:
Isn't that just moving things up one level?  Since the contents of hal_nxp/CMakeLists.txt are adding subdirs there?

Reply:
@carlescufi I think this is how the NRF 802.15.4 drivers are also integrated into the build:

https://github.com/zephyrproject-rtos/zephyr/blob/58b8c4006e1bc0bb2ee8056b807ecb11e1be71bb/modules/hal_nordic/nrf_802154/CMakeLists.txt#L103

Reply:
Sure, but the idea is to put any Zephyr-specific CMake code in the main zephyr tree, and leave RTOS-neutral CMake in the module repo.

Reply:
@carlescufi sure for certain modules I get how that makes sense.  For the NXP HAL I was trying to start with transition zephyr specific glue code into the Zephyr tree and leave the build system changes as is.

Reply:
> 
> 
> @carlescufi sure for certain modules I get how that makes sense. For the NXP HAL I was trying to start with transition zephyr specific glue code into the Zephyr tree and leave the build system changes as is.

Looking at the CMake code here https://github.com/zephyrproject-rtos/hal_nxp/, then in practice all CMake code down the tree is Zephyr specific, for example:
https://github.com/zephyrproject-rtos/hal_nxp/blob/1d823ea49bf7f6229a1097fbff33e27ba0dd389b/mcux/drivers/kinetis/CMakeLists.txt#L14
uses Zephyr specific CMake functions:
```
zephyr_library_sources_ifdef(CONFIG_ADC_MCUX_ADC12	fsl_adc12.c)
....
```

and changing that to be reusable outside Zephyr is a much bigger task, not being handled in this PR.

However, if the CMake code in hal_nxp was made split in Zephyr glue, and pure CMake, then moving the Zephyr specific code into Zephyr repo would make good sense.

But as this Pr stands now, I don't see the gain in calling https://github.com/zephyrproject-rtos/hal_nxp/blob/master/CMakeLists.txt
from here, as
```
add_subdirectory(${ZEPHYR_CURRENT_MODULE_DIR} hal_nxp)
```
I think it just makes it harder follow how hal_nxp gets integrated into Zephyr build system.

As @carlescufi  referenced, the hal_nordic only contains pure CMake code, with no dependencies to Zephyr, for example:
https://github.com/zephyrproject-rtos/hal_nordic/blob/master/drivers/nrf_802154/CMakeLists.txt
https://github.com/zephyrproject-rtos/hal_nordic/blob/master/drivers/nrf_802154/driver/CMakeLists.txt

Reply:
@tejlmand how would you suggest we move zephyr glue code into zephyr but leave build systems alone that might be zephyr based?

Reply:
This is a bit of a special case IMHO. The CMake files are all Zephyr-specific but currently are in the module instead of being in the main tree. It might make sense to move all CMakeLists files to zephyr/modules/hal_nxp and keep a pristine copy of mcux in the repo? @MaureenHelm should also weigh in here.

Reply:
>For the NXP HAL I was trying to start with transition zephyr specific glue code into the Zephyr tree

Can you point us to the glue code?

>how would you suggest we move zephyr glue code into zephyr but leave build systems alone that might be zephyr based?

So, you want to move the **code** to the main tree but leave the zephyr-specific CMake files in the external repo, right? But you of course also need to move the CMake files that deal with the glue code into the main zephyr tree. Maybe if you point us to that glue code and how it is being built today we can take a closer look.

Reply:
For the NXP case, you can see the move of the USB glue code here:

https://github.com/zephyrproject-rtos/zephyr/pull/37215/files

We have a similar case for the TI HAL (no changes made yet, but it has glue code you can see here: https://github.com/zephyrproject-rtos/hal_ti/tree/master/simplelink/kernel/zephyr/dpl).